### PR TITLE
[REF] refactor FLM fitting

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes none
- Extracted from #4112

Changes proposed in this pull request:

- refactor fitting of first level models into several smaller functions

### CODE REVIEW

**Review by bthirion — COMMENTED** [review]
I Like thre refactoring. I'm still wondering whether we want to support throwing ndarrays into the GLM: people will do crazy things and get back to us...

File: nilearn/glm/first_level/first_level.py [hunk 1]
```diff
@@ -490,6 +489,210 @@ def scaling_axis(self):
...
+            return design_matrices
+
+        design_matrices = []
+
+        for run_idx, run_img in enumerate(run_imgs):
+
+            if isinstance(run_img, SurfaceImage):
+                n_scans = run_img.shape[0]
+            elif isinstance(run_img, np.ndarray):
+                n_scans = run_img.shape[1]
```
Comment (Remi-Gau) [inline nilearn/glm/first_level/first_level.py:650]: remove

File: nilearn/glm/first_level/first_level.py [hunk 2]
```diff
@@ -743,10 +841,14 @@ def compute_contrast(
...
 
+        format_type : str, default='niimg'
+            Data format type of the output. Can be ``'niimg'``, or ``'array'``.
+            If ``'niimg'`` is chosen a SurfaceImage or NiftiImage instance
+            will be returned depending on the type of the GLM inputs.
 
         Returns
         -------
-        output : Nifti1Image or dict
+        output : Nifti1Image, numpy array, or dict
```
Comment (Remi-Gau) [inline nilearn/glm/first_level/first_level.py:851]: remove

File: nilearn/glm/first_level/first_level.py [hunk 3]
```diff
@@ -490,6 +489,210 @@ def scaling_axis(self):
...
+    def _fit_single_run(self, sample_masks, bins, run_img, run_idx):
+        """Fit the model for a single and keep only the regression results."""
+        design = self.design_matrices_[run_idx]
+
+        sample_mask = None
+        if sample_masks is not None:
+            sample_mask = sample_masks[run_idx]
+            design = design.iloc[sample_mask, :]
+            self.design_matrices_[run_idx] = design
+            if isinstance(run_img, np.ndarray):
```
Comment (Remi-Gau) [inline nilearn/glm/first_level/first_level.py:594]: remove

**Comment by Remi-Gau** [comment]
> whether we want to support throwing ndarrays into the GLM

my bad I had forgotten a few lines that still related to ndarrays: should look better now.

**Review by bthirion — APPROVED** [review]
The changes LGTM. Thx !

**Pull request merged**